### PR TITLE
Vite: Force vite to always optimize Scrivito again.

### PR DESCRIPTION
**Pull request by apepper**
Otherwise vite will sometimes fail with the following js console error:

```
Failed to load resource: the server responded with a status of 504 (Outdated Optimize Dep)
```

See https://github.com/vitejs/vite/issues/12434 for the inspiration (using `exclude` instead of `include`).

My usual workaround for this was to `rm -rf node_modules/.vite` if I see this. But I don't want to tell this to every dev using the portal app.

### CODE REVIEW

File: vite.config.ts [hunk 1]
```diff
@@ -25,6 +25,10 @@ export default defineConfig(({ mode }) => {
       'import.meta.env.ENABLE_NEOLETTER_FORM_BUILDER_SUBSCRIPTION_FEATURE':
         JSON.stringify(env.ENABLE_NEOLETTER_FORM_BUILDER_SUBSCRIPTION_FEATURE),
     },
+    optimizeDeps: {
+      include: ['scrivito'],
+      force: true,
```
[the remark this answers is not in the document]
Comment (apepper) [reply vite.config.ts:30]: Yes, but with the `force: true` I'm telling vite to always optimize `scrivito` again, see https://vitejs.dev/config/dep-optimization-options.html#optimizedeps-force.

[the remark this answers is not in the document]
Comment (dcsaszar) [reply vite.config.ts:30]: With the docs being as bad as they are and no additional info regarding the `include/force` combo online, and no way of reproducing the issue, unfortunately I can't review this.
I see that [newer vice versions](https://github.com/vitejs/vite/blob/main/packages/vite/CHANGELOG.md#4410-2023-10-03) have at least one "Outdated Optimize Dep"-related fix ([14284](https://github.com/vitejs/vite/issues/14284)).

Is there some sensible way to reproduce this?

**Comment by apepper** [comment]
> Is there some sensible way to reproduce this?

I don't have a reliable way to reproduce it. The closes I got so far is this:

* `git checkout main`
* `npm i && npm start`
* Visit http://localhost:8080 in your browser - the Scrivito UI opens up and the portal app is visible.
* Leave the tab open
* Now switch to a different branch _while still leaving `npm start` running_! E.g. https://github.com/apepper/scrivito-portal-app/tree/hackathon-rest-in-peace
* Look in your browser again - the page is still visible
* Now stop `npm start` and run `npm i && npm start` again. Look in your browser.

_Expected_
The portal app renders correctly

_Actual_

Sometimes  - not always - the portal app continues to show a spinner and the js console shows:

```
Failed to load resource: the server responded with a status of 504 (Outdated Optimize Dep)
```

**Comment by dcsaszar** [comment]
I ended up with a test env switching branches every 5 seconds, constantly running npm install, and manually stopping and starting the server.
* I tested with and without UI (the repro is ambiguous in this regard)
* It crashed vite more than 10 times (needs killing the node process).
* I saw some "Outdated Optimize Dep", e.g.:
  ```ruby
  GET http://localhost:8080/node_modules/.vite/deps/react-helmet-async.js?v=bb5b79b7 net::ERR_ABORTED 504 (Outdated Optimize Dep)
  GET http://localhost:8080/node_modules/.vite/deps/react_jsx-dev-runtime.js?v=e5f3a145 504 (Outdated Optimize Dep)
  GET http://localhost:8080/node_modules/.vite/deps/react.js?v=5b2bd4cb net::ERR_ABORTED 504 (Outdated Optimize Dep)
  GET http://localhost:8080/node_modules/.vite/deps/react-dom_client.js?v=882440fd net::ERR_ABORTED 504 (Outdated Optimize Dep)
  GET http://localhost:8080/node_modules/.vite/deps/scrivito.js?v=55fb69b1 net::ERR_ABORTED 504 (Outdated Optimize Dep)
  ```
* The app was never permanently in the spinning state. If the browser showed the spinner (or was otherwise no longer updating), a browser reload fixed it.

I'm afraid we need more info on how to reproduce the issue.

**Comment by apepper** [comment]
> I'm afraid we need more info on how to reproduce the issue.

Did you use Chrome? I tried to reproduce my steps with Chrome but wasn't able to. With Brave I can reproduce it quite often.

**Comment by dcsaszar** [comment]
> > I'm afraid we need more info on how to reproduce the issue.
> 
> Did you use Chrome? I tried to reproduce my steps with Chrome but wasn't able to. With Brave I can reproduce it quite often.

I tried my setup with Brave Version 1.60.118 Chromium: 119.0.6045.163 (arm64) and the results are the same as with my Chrome: Can't reproduce locally.

**Comment by apepper** [comment]
Closing since vite is now at version 5 (with #223). Let's see if this issue reappears.

File: vite.config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcsaszar) [reply vite.config.ts:30]: FYI I currently can reproduce the bug when switching between these two branches:
* `portal_only_mode`
* `dave-content-restore`

Putting `optimizeDeps: { force: true },` fixes the issue for me.

[the remark this answers is not in the document]
Comment (dcsaszar) [reply vite.config.ts:34]: * [ ] I still don't understand why we need `include: ['scrivito'],`.

[the remark this answers is not in the document]
Comment (apepper) [reply vite.config.ts:34]: I haven't checked this, but https://vitejs.dev/config/dep-optimization-options.html#optimizedeps-force says: 

> optimizeDeps.force
> Type: boolean
> Set to `true` to force dependency pre-bundling, ignoring previously cached optimized dependencies.

So if we don't limit the scope to `include: ['scrivito']` _all_ packages will be force build, not just `scrivito`.

[the remark this answers is not in the document]
Comment (dcsaszar) [reply vite.config.ts:34]: My understanding of the docs is different:
> optimizeDeps.include
> By default, linked packages not inside node_modules are not pre-bundled. Use this option to force a linked package to be pre-bundled.

I read this as an option exclusively meant for packages **in addition to the default**.

The function name `addManuallyIncludedOptimizeDeps` supports this argument. https://github.com/vitejs/vite/blob/0cacfadf3aeb3c9ce1db490290584387efb4f5ef/packages/vite/src/node/optimizer/index.ts#L253

**Comment by apepper** [comment]
FEEDBACK ([a25df02](https://github.com/Scrivito/scrivito-portal-app/pull/181/commits/a25df024ba5a10b834722d0466d6764bfcbbfa54))

**Pull request merged**